ocaml/tests/test_clustering.ml

```diff
@@ -376,12 +376,205 @@ let test_disallow_unplug_ro_with_clustering_enabled =
+ * If any SRs of matching type are found, the first corresponding cluster_stack is returned,
+ * If no SRs are found it returns the existing cluster's stack, otherwise it returns the
+ * default cluster_stack in Xapi_globs *)
+
+let test_choose_cluster_stack_clusters_no_sms () =
+  let __context = T.make_test_database () in
+  choose_cluster_stack_should_select default ~__context;
+
+  (* Add two cluster, test choose_cluster_stack's filtering *)
+  for i = 0 to 1 do
+    let _ = T.make_cluster_and_cluster_host ~__context () in
+    choose_cluster_stack_should_select default_smapiv3 ~__context
+  done;
+
+  (* choose_cluster_stack should fail with multiple different cluster_stacks *)
+  let _ = T.make_cluster_and_cluster_host ~__context ~cluster_stack:"different_stack" () in
+  choose_cluster_stack_should_fail_with_conflict ~__context
+
+let create_and_plug_sr ~_type ~__context () =
+  let host = Helpers.get_localhost ~__context in (* plug to master *)
+  let sR : API.ref_SR = T.make_sr ~__context ~_type () in
+  let _  : API.ref_PBD = T.make_pbd ~__context ~host ~currently_attached:true ~sR () in
+  sR
+
+let test_choose_cluster_stack_sms_no_clusters () =
+  let __context = T.make_test_database () in
+
+  (* create SMs with cluster_stack constraints *)
+  let _  : API.ref_SM = T.make_sm ~__context ~_type:"sm" ~required_cluster_stack:[ test_default ] () in
+  let _  : API.ref_SM = T.make_sm ~__context ~_type:"nfs" ~required_cluster_stack:[] () in
+  let _  : API.ref_SM = T.make_sm ~__context ~_type:"lvm" ~required_cluster_stack:[ default_smapiv3 ; default ] () in
+
+  let sm_sr = create_and_plug_sr ~_type:"sm" ~__context () in
+  choose_cluster_stack_should_select test_default ~__context;
+
+  (* New SR doesn't add additional constraints as NFS SM has no required_cluster_stacks *)
+  let _ = create_and_plug_sr ~_type:"nfs" ~__context () in
+  choose_cluster_stack_should_select test_default ~__context;
+
+  (* No common required cluster stack between LVM and SM types *)
+  let _ = create_and_plug_sr ~_type:"lvm" ~__context () in
+  choose_cluster_stack_should_fail_with_conflict ~__context;
+
+  (* Remove conflict, now first LVM stack will be selected *)
+  begin match Db.SR.get_PBDs ~__context ~self:sm_sr with
+  | [ pBD ] ->
+    Db.PBD.set_currently_attached ~__context ~self:pBD ~value:false;
+    Db.SR.destroy ~__context ~self:sm_sr
+  | _ -> Alcotest.fail "only one PBD should be plugged into this SR"
+  end;
+  choose_cluster_stack_should_select default_smapiv3 ~__context;
+
+  (* default_smapiv3 is the only common cluster stack *)
+  let _  : API.ref_SM = T.make_sm ~__context ~_type:"hba" ~required_cluster_stack:[ test_default ; default_smapiv3 ] () in
+  let _ = create_and_plug_sr ~_type:"hba" ~__context () in
+  choose_cluster_stack_should_select default_smapiv3 ~__context;
+
+  let _  : API.ref_SM = T.make_sm ~__context ~_type:"gfs2" ~required_cluster_stack:[ default_smapiv3 ; default_smapiv3 ] () in
```

Comment:
did you mean to repeat the stack twice here?

Reply:
Yes, I wanted to check that doesn't confuse `choose_cluster_stacks`.